modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/processing/BlockAcceptanceManager.scala

```diff
@@ -0,0 +1,92 @@
+package org.tessellation.infrastructure.snapshot.processing
+
+import cats.Applicative
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.eq._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.show._
+
+import org.tessellation.dag.domain.block.{BlockReference, DAGBlock}
+import org.tessellation.dag.snapshot.BlockAsActiveTip
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signed
+
+import monocle.syntax.all._
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait BlockAcceptanceManager[F[_]] {
+
+  def processBlocks(
+    initState: BlockAcceptanceState,
+    blocks: List[Signed[DAGBlock]]
+  ): F[BlockAcceptanceResult]
+
+}
+
+object BlockAcceptanceManager {
+
+  def make[F[_]: Async: KryoSerializer: SecurityProvider](logic: BlockAcceptanceLogic[F]): BlockAcceptanceManager[F] =
+    new BlockAcceptanceManager[F] {
+      private val logger = Slf4jLogger.getLoggerFromClass[F](BlockAcceptanceManager.getClass)
+
+      def processBlocks(
+        initState: BlockAcceptanceState,
+        blocks: List[Signed[DAGBlock]]
+      ): F[BlockAcceptanceResult] = {
+
+        def go(prevResult: BlockAcceptanceResult): F[BlockAcceptanceResult] = {
```

Comment:
can we make it tail recursive?

Reply:
We can but there's no need to. You can have a recursive function of arbitrary depth with `IO[_]` returned from `flatMap`. Also the code without tail-recursion is easier to read.